Improve Binary Search

glebkema opened the pull request:
I didn't сombine the commits because I used them to group and explain my changes. If necessary, I will redo the PR.

1. toughen up tests
    - check more missing and present values
    - check triple value search (and let the second group of tests not repeat the first one)
    - check search for null in empty array (to make sure the algorithm distinguishes zero from emptiness)
2. optimize the algorithm
    - use `return` to stop the loop and simplify the `if then` structure
    - sort the array on initialization, not before each search

- [x] tests passed successfully
- [x] maven `compile` executed

glebkema committed:
BinarySearchTest / check all values close to array elements

glebkema committed:
BinarySearchTest / check for a triple repeat of a value instead of duplicating the first test

glebkema committed:
BinarySearchTest / check search for null in empty array

glebkema committed:
BinarySearch / simplify the search logic

glebkema committed:
BinarySearch / sort the array on initialization, not before each search

Pradyuman7 commented on src/main/java/com/pd/algorithms/BinarySearch.java:
```diff
@@ -8,28 +8,28 @@ public class BinarySearch {
     return binarySearch(arr, 0, size - 1, num);
   }
 
   public boolean binarySearch(int[] array, int low, int high, int value) {
```
I guess we can make this method `private` since this is mainly to be called from the `search` public method. Could you please update this and also update the docs, if needed?

glebkema commented:
Ok, I will do it in about 12 hours

Pradyuman Dixit ***@***.***> schrieb am Di., 11. Okt. 2022,
09:20:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/main/java/com/pd/algorithms/BinarySearch.java
> <https://github.com/Pradyuman7/Algorithmics/pull/55#discussion_r991903215>
> :
>
> >      this.arr = arr;
>      this.size = arr.length;
>    }
>
>    public boolean search(int num) {
> -    Arrays.sort(arr);
>      return binarySearch(arr, 0, size - 1, num);
>    }
>
>    public boolean binarySearch(int[] array, int low, int high, int value) {
>
> I guess we can make this method private since this is mainly to be called
> from the search public method. Could you please update this and also
> update the docs, if needed?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Pradyuman7/Algorithmics/pull/55#pullrequestreview-1136891816>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA5TUYH6YZ2CWWAZAWGQFZLWCUIMRANCNFSM6AAAAAARBU7YGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

glebkema committed:
use binarySearch() as private recursive method

glebkema committed:
BinarySearch / remove loop from recursion

glebkema commented:
I realized that our recursion does not require a `while` loop. So I replaced it with an `if` statement and reordered all the `if` statements to make the code clearer. 

Also I added some more tests for negative values and zero in the array.

Pradyuman7 reviewed (approved):
LGTM 💯
